### PR TITLE
CalDAV Integration

### LINKED ISSUE #24: Import timesheets from agreed external systems
- [ ] iCal/CalDAV

* ~~Scoro~~
* ~~Mite~~
* ~~Time~~

(not needed since these are current [Tiki integrations](https://gitter.im/federatedbookkeeping/timesheets?at=62ab2467e393a318065d4dd1))

See https://github.com/federatedbookkeeping/timesheets#milestone-4-documentation

note: https://gitter.im/federatedbookkeeping/timesheets?at=62bc3bd49a314a3ec4f82ba1

- [ ] If additionally integrating with other time-tracking systems, any data received from them must also be shared with federated systems.

### LINKED ISSUE #71: Per-instance integration configuration
Integration configuration can only be injected via environment variables.

Should be possible to dd configuration per integration instance. Essential for multi-tenant and multi-project integrations, e.g. for `TIMELD_GATEWAY_PREJOURNAL__USER`, `TIMELD_GATEWAY_PREJOURNAL__CLIENT`

### PR DESCRIPTION
**Pull request by gsvarovsky**
Closes #71: Instance-specific integration configuration
Closes #24: Importing timesheet entries from CalDAV

### CODE REVIEW

**Comment by gsvarovsky** [comment]
@Angus-McAllister : READMEs to review in this PR:
- packages/cli/README.md (note, links to not yet published npm repos)
- packages/caldav/README.md
- packages/prejournal/README.md
- packages/tiki/README.md

File: packages/caldav/README.md [hunk 1]
```diff
@@ -0,0 +1,33 @@
+# timeld - CalDAV integration
+
+This packages supports live update of timesheet entries:
```
Comment (Angus-McAllister) [inline packages/caldav/README.md:3]: Should be "package" singular, rather than plural

File: packages/caldav/README.md [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+- _from_ a CalDAV end-point
```
Comment (Angus-McAllister) [inline packages/caldav/README.md:4]: I'd suggest adding an example here along the lines of "such as iCloud Calendar".

File: packages/caldav/README.md [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+- _to_ **timeld**.
+
+## installation & configuration
+
+This integration package is included in the gateway.
```
Comment (Angus-McAllister) [inline packages/caldav/README.md:9]: I'd suggest capitalising "Gateway", to ensure readers know the documentation is referring to a specific timeld component.

File: packages/caldav/README.md [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+The integration requires the following configuration:
+- `uri`: calendar URI (https)
```
Comment (Angus-McAllister) [inline packages/caldav/README.md:12]: `url`, surely?  Rationale: it has to be reachable, so requires a concrete reference.

File: packages/caldav/README.md [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+- `auth.user`: CalDAV account user (optional if calendar is public)
+- `auth.pass`: CalDAV account password (optional if calendar is public)
+- `owner`: calendar owner IRI (optional; defaults to the timesheet owner account)
+- `pollInterval` CalDAV polling interval, in millis (optional; defaults to 10000)
```
Comment (Angus-McAllister) [inline packages/caldav/README.md:16]: For the time unit, either use the SI abbreviation "ms" or express it in full as "milliseconds".

File: packages/caldav/README.md [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+Configuration defaults may already be included in the gateway. Timesheet-specific configuration can also be included when adding the integration, see below.
```
Comment (Angus-McAllister) [inline packages/caldav/README.md:18]: 1. Ditto re: case of "Gateway".
2. As per conversation, replace "integration" with "connector".

File: packages/caldav/README.md [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+To activate the integration for a timesheet "my-work" (note this assumes authorisation has defaults in the gateway).
```
Comment (Angus-McAllister) [inline packages/caldav/README.md:20]: Ditto re: case of "Gateway"

File: packages/caldav/README.md [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+```shell
+timeld admin
+alice> add integration timeld-caldav --ts my-work --config.uri "https://p116-caldav.icloud.com/151157399/calendars/work"
```
Comment (Angus-McAllister) [inline packages/caldav/README.md:24]: As per earlier conversation, replace `integration` with `connector`.

File: README.md [hunk 9]
```diff
@@ -25,14 +26,12 @@ The [desktop tool](https://www.npmjs.com/package/timeld-cli) (called **timeld-cl
 The [Gateway](https://www.npmjs.com/package/timeld-gateway) manages accounts and persists timesheets somewhere safe.
 
 - Register accounts
-- ![coming soon](https://img.shields.io/badge/-coming%20soon-red) Manage projects
-- ![coming soon](https://img.shields.io/badge/-coming%20soon-red) Submit completed timesheets to other management systems
-- ![coming soon](https://img.shields.io/badge/-coming%20soon-red) Import timesheets from other management systems
+- Manage projects
+- Submit completed timesheets to other management systems
```
Comment (Angus-McAllister) [inline README.md:30]: I'd suggest the term "time-tracking systems" in preference to "management systems".

File: README.md [hunk 10]
```diff
@@ -25,14 +26,12 @@ The [desktop tool](https://www.npmjs.com/package/timeld-cli) (called **timeld-cl
...
+- Import timesheets from other management systems
```
Comment (Angus-McAllister) [inline README.md:31]: Ditto re: "time-tracking systems"

File: packages/cli/README.md [hunk 11]
```diff
@@ -118,12 +118,22 @@ The output of all the `list` and `report` commands in a session can be piped to
...
+|------------------------------------------------------------------------------|----------------------------------------------------------------------|
+| Ponder Source's [PreJournal](https://github.com/pondersource/prejournal/)    | [timeld-prejournal](https://www.npmjs.com/package/timeld-prejournal) |
+| Evoludata's [Tiki](https://timesheet.dev3.evoludata.com/Timesheets-homepage) | [timeld-tiki](https://www.npmjs.com/package/timeld-tiki)             |
+| [CalDAV](https://www.rfc-editor.org/rfc/rfc4791)                             | [timeld-caldav](https://www.npmjs.com/package/timeld-caldav)         |
+
+This is done individually for each timesheet; here's how:
 
 ```
-add integration integration-module --timesheet my-timesheet
+add integration ≪integration-module≫ --timesheet my-timesheet --config.≪key≫ ≪config-value≫
```
Comment (Angus-McAllister) [inline packages/cli/README.md:132]: As per earlier conversation, replace `integration` with `connector` in each case.

File: packages/cli/README.md [hunk 12]
```diff
@@ -118,12 +118,22 @@ The output of all the `list` and `report` commands in a session can be piped to
 report my-project --format json-ld > my-project-report.json
 ```
 
-You can synchronise your timesheets with other federated time-tracking systems (currently Ponder Source's [PreJournal](https://github.com/pondersource/prejournal/) and Evoludata's [Tiki](https://timesheet.dev3.evoludata.com/Timesheets-homepage)).  This is done individually for each timesheet; here's how:
+You can synchronise your timesheets with other federated time-tracking systems; currently:
+
+| system                                                                       | integration module                                                   |
```
Comment (Angus-McAllister) [inline packages/cli/README.md:123]: Nice improvement to the formatting.

File: packages/cli/README.md [hunk 13]
```diff
@@ -118,12 +118,22 @@ The output of all the `list` and `report` commands in a session can be piped to
...
+
+This is done individually for each timesheet; here's how:
 
 ```
-add integration integration-module --timesheet my-timesheet
+add integration ≪integration-module≫ --timesheet my-timesheet --config.≪key≫ ≪config-value≫
 ```
-where `integration-module` is either `timeld-prejournal` or `timeld-tiki`.
+where:
+- `≪integration-module≫` is `timeld-prejournal`, `timeld-tiki`, or `timeld-caldav`
```
Comment (Angus-McAllister) [inline packages/cli/README.md:135]: 1. Ditto re: `connector`
2. I'd suggest "is one of `timeld-prejournal,`" etc.

File: packages/cli/README.md [hunk 14]
```diff
@@ -118,12 +118,22 @@ The output of all the `list` and `report` commands in a session can be piped to
...
+- `≪key≫` and `≪config-value≫` are any configuration required for the integration (see integration module links above)
```
Comment (Angus-McAllister) [inline packages/cli/README.md:136]: Ditto re: `connector`

File: packages/prejournal/README.md [hunk 15]
```diff
@@ -1,6 +1,29 @@
-⚠️ This package is an untested sketch for [prejournal](https://prejournal.org/) timesheets  sync integration.
+# timeld - Prejournal integration
```
Comment (Angus-McAllister) [inline packages/prejournal/README.md:1]: In view of earlier conversation, would suggest "PreJournal Connector for timeld".

File: packages/prejournal/README.md [hunk 16]
```diff
@@ -1,6 +1,29 @@
...
 
-## API docs
+This packages supports live update of timesheet entries:
```
Comment (Angus-McAllister) [inline packages/prejournal/README.md:3]: "packages" should be singular.

File: packages/prejournal/README.md [hunk 17]
```diff
@@ -1,6 +1,29 @@
...
+- _from_ **timeld**
+- _to_ [prejournal](https://prejournal.org/).
+
+## installation & configuration
+
+This integration package is included in the gateway.
```
Comment (Angus-McAllister) [inline packages/prejournal/README.md:9]: As per earlier conversation, replace "integration package" with "connector".

File: packages/prejournal/README.md [hunk 18]
```diff
@@ -1,6 +1,29 @@
...
+
+The integration requires the following configuration:
```
Comment (Angus-McAllister) [inline packages/prejournal/README.md:11]: Ditto re: "connector"

File: packages/prejournal/README.md [hunk 19]
```diff
@@ -1,6 +1,29 @@
...
+- `user`: Prejournal account (user) name (required)
+- `key`: Prejournal account key (required)
+- `api`: API URL e.g. `https://time.pondersource.com/v1/` (required)
+- `client`: default timesheet "client" (for whom the work is being done; required)
+
+Configuration defaults may already be included in the gateway. Timesheet-specific configuration can also be included when adding the integration, see below.
```
Comment (Angus-McAllister) [inline packages/prejournal/README.md:17]: 1. As mentioned in earlier comment on another README.md in this review, would recommend capitalising "Gateway"; and
2. Replace "integration" with "connector".

File: packages/prejournal/README.md [hunk 20]
```diff
@@ -1,6 +1,29 @@
...
+
+To activate the integration for a timesheet "my-work" (note this assumes authorisation and API defaults in the gateway).
```
Comment (Angus-McAllister) [inline packages/prejournal/README.md:19]: Ditto re: "connector".

File: packages/tiki/README.md [hunk 21]
```diff
@@ -1,6 +1,27 @@
...
+- `api`: Tiki tracker items API URL e.g. `https://timesheet.dev3.evoludata.com/api/trackers/2/items`
+- `token`: Tiki OAuth2 token
+
+Configuration defaults may already be included in the gateway. Timesheet-specific configuration can also be included when adding the integration, see below.
+
+To activate the integration for a timesheet "my-work" (note this assumes authorisation has defaults in the gateway).
+
+```shell
+timeld admin
+alice> add integration timeld-tiki --ts my-work --config.api "https://timesheet.dev3.evoludata.com/api/trackers/2/items"
```
Comment (Angus-McAllister) [inline packages/tiki/README.md:21]: Ditto re: `connector`.

**Review by Angus-McAllister — CHANGES_REQUESTED** [review]
Most comments relate to our earlier conversation about disambiguating the term "integration" by introducing "connector".

**Review by Angus-McAllister — APPROVED** [review]
Perfect, thx :-)

**Pull request merged**